Source/WebCore/platform/graphics/cocoa/H264UtilitiesCocoa.mm

```diff
@@ -0,0 +1,123 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "config.h"
+#import "H264UtilitiesCocoa.h"
+
+#import "BitReader.h"
+#import "MediaSample.h"
+
+#import <pal/cf/CoreMediaSoftLink.h>
+
+namespace WebCore {
+
+RefPtr<VideoInfo> createVideoInfoFromAVCC(std::span<const uint8_t> avcc)
+{
+    if (avcc.size() < 7)
+        return nullptr;
+
+    BitReader reader { avcc };
+
+    // configurationVersion
+    reader.read(8);
+    // AVCProfileIndication;
+    reader.read(8);
+    // profile_compatibility;
+    reader.read(8);
+    // AVCLevelIndication;
+    reader.read(8);
+    // bit(6) reserved = '111111'b;
+    // unsigned int(2) lengthSizeMinusOne;
+    size_t lengthSize = (*reader.read<uint8_t>() & 0x3) + 1;
+    // bit(3) reserved = '111'b;
+    // unsigned int(5) numOfSequenceParameterSets;
+    size_t numOfSequenceParameterSets = 0x1f & *reader.read<uint8_t>();
+    if (!numOfSequenceParameterSets)
+        return nullptr;
+
+    constexpr size_t kNALTypeSize = 1;
+    constexpr uint8_t kSPSNAL = 7;
+    constexpr uint8_t kPPSNAL = 8;
+
+    Vector<const uint8_t*> paramSetPtrs;
+    paramSetPtrs.reserveInitialCapacity(numOfSequenceParameterSets + 1);
+    Vector<size_t> paramSetSizes;
+    paramSetSizes.reserveInitialCapacity(numOfSequenceParameterSets + 1);
+    size_t offset = 6;
+    for (size_t index = 0; index < numOfSequenceParameterSets; index++) {
+        auto size = reader.read<uint16_t>();
+        if (!size || *size < kNALTypeSize)
+            return nullptr;
+        offset += 2;
+        auto nalType = reader.read<uint8_t>();
+        if (!nalType || (*nalType & 0x1f) != kSPSNAL)
+            return nullptr;
+        const_cast<uint8_t*>(avcc.subspan(offset).data())[0] &= 0x7f;
+        const_cast<uint8_t*>(avcc.subspan(offset).data())[0] |= 0x60;
```

Comment:
This is rather odd to modify in place the avcc input parameter.
Can we instead copy the data somewhere else?
And also, can we add names to 0x7f, for instance naluTypeMask? Ditto for 0x60